pkg/io/opener.go

```diff
@@ -208,3 +210,86 @@ func (o *opener) Writer(ctx context.Context, path string) (io.WriteCloser, error
 	}
 	return writer, nil
 }
+
+type UploadFunc func() (io.Reader, *blob.Attributes, error)
+
+func FileUpload(file string, attr *blob.Attributes) UploadFunc {
+	return func() (io.Reader, *blob.Attributes, error) {
+		reader, err := os.Open(file)
+		if err != nil {
+			return nil, nil, err
+		}
+		return reader, attr, nil
+	}
+}
+
+func DataUpload(reader io.Reader, attr *blob.Attributes) UploadFunc {
+	return func() (io.Reader, *blob.Attributes, error) {
+		return reader, attr, nil
+	}
+}
+
+func (o *opener) Upload(ctx context.Context, uploads map[string]UploadFunc) error {
+	errCh := make(chan error, len(uploads))
+	group := &sync.WaitGroup{}
+	group.Add(len(uploads))
+	for dest, upload := range uploads {
+		log := logrus.WithField("dest", dest)
+		log.Infof("Queued %s for upload", dest)
+		go func(f UploadFunc, dest string, log *logrus.Entry) {
+			defer group.Done()
+			//reader, attr, err := f()
+			reader, _, err := f()
+			if err != nil {
+				errCh <- err
+			}
+
+			// FIXME
+			//writer, err := o.Writer(ctx, dest, convertAttributesToWriterOptions(attr))
+			writer, err := o.Writer(ctx, dest)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			_, copyErr := io.Copy(writer, reader)
+			if copyErr != nil {
+				copyErr = fmt.Errorf("copy error: %v", copyErr)
+			}
+			closeErr := writer.Close()
+			if closeErr != nil {
+				closeErr = fmt.Errorf("writer close error: %v", closeErr)
+			}
+			err = errorutil.NewAggregate(copyErr, closeErr)
+			if err != nil {
+				errCh <- err
+			} else {
+				log.Info("Finished upload")
+			}
+		}(upload, dest, log)
+	}
+	group.Wait()
+	close(errCh)
+	if len(errCh) != 0 {
+		var uploadErrors []error
+		for err := range errCh {
+			uploadErrors = append(uploadErrors, err)
+		}
+		return fmt.Errorf("encountered errors during upload: %v", uploadErrors)
+	}
+	return nil
+}
+
+func convertAttributesToWriterOptions(a *blob.Attributes) *blob.WriterOptions {
+	if a == nil {
+		return nil
+	}
+	return &blob.WriterOptions{
+		CacheControl:       a.CacheControl,
```

Comment:
Which of these do we actually use?

Reply:
Not used in the current version. That's because of the rebase. Depending on how/where we want to implement the upload functionality I would either keep it or move it away. (It was actually used in l.248. but that's not possible anymore because `Writer(...)` doesn't have a WriterOptions parameter anymore. I would clean this up of course as soon as we decide where the Upload functionality should be implemented.